### ISSUE TITLE
Enhancement: no-meaningless-void-operator should allow `void` if the expected assigned type is `undefined`

### ISSUE THREAD

**Issue by Sxxov** [typescript-eslint/typescript-eslint#7358]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-meaningless-void-operator

## Explanation

Currently, TypeScript will not allow the return type of a `void` function to be used to assign to a variable that allows `undefined`, [by design](https://github.com/Microsoft/TypeScript/issues/25481).

```ts
let foo: number | undefined;
const voidFn = (): void => {};

foo = voidFn();
// Type 'void' is not assignable to type 'number | undefined'. ts(2322)
```

In cases where the `no-meaningless-void-operator` rule is active, & a function returns anything other than `undefined`, we may use the `void` operator before it to discard the value & assign `undefined` to a variable.

```ts
let foo: number | undefined;
const strFn = (): string => 'bar';

foo = void strFn(); // ok
```

However, the `no-meaningless-void-operator` rule currently forbids discarding results of `void` functions, even if the returned result enables us to assign to a variable where TypeScript expects to be `undefined`.

```ts
let foo: number | undefined;
const voidFn = (): void => {};

foo = void voidFn();
// void operator shouldn't be used on void; it should convey that a return value is being ignored eslint(@typescript-eslint/no-meaningless-void-operator)
```

## Use-case

Basically anywhere you would want to clear a variable when you call a cleanup function. For example, using `cancelAnimationFrame` that clears the handle variable & cancels the scheduled task:

```ts
rafHandle = void cancelAnimationFrame(rafHandle);
```

## Related

[`no-confusing-void-expression`](https://typescript-eslint.io/rules/no-confusing-void-expression): The above described patterns still don't work even with the `ignoreVoidOperator` option enabled.

### Fail

```typescript
let foo: number | undefined;
const voidFn = (): void => {};

foo = voidFn();
```

### Pass

```typescript
let foo: number | undefined;
const voidFn = (): void => {};

foo = void voidFn();
```

**Comment by bradzacher** [comment]
I'm not quite sure I understand why you're using the `void` operator for this case instead of, say, assigning `undefined`?
```ts
voidFn();
foo = undefined;
```

What is your usecase? Is it just for terseness?

**Comment by Sxxov** [comment]
> I'm not quite sure I understand why you're using the `void` operator for this case instead of, say, assigning `undefined`?
> 
> ```ts
> 
> voidFn();
> 
> foo = undefined;
> 
> ```
> 
> 
> 
> What is your usecase? Is it just for terseness?
> 
> 
> 
> As written I just don't see the broader applicability of this. The rule itself is already pretty niche and this style of TS is even nicher.

For the use case, basically, yeah. However, the main motivation is that it feels like a thorn on the side that it _doesn't_ work, especially because of all the other rules & config around the void operator.

I think it fits the same argument as allowing it for arrow functions, where you don't want extra parentheses. Most notably, this isn't a zero-occurrence snag in my Svelte codebase (with the `$:` operator).

With all that said though, I do agree that it's pretty niche, & I would 100% understand if it's a `won't-fix` due to scope, but man it just doesn't feel right.

**Comment by bradzacher** [comment]
I disagree that it's similar to the arrow function case in `no-confusing-void-expression`.
That case makes sense as an option because it's idiomatic JS to write code like `() => returnsVoid()`. Expanding that code to `() => { returnsVoid() }` does do the same thing, for sure - except this clashes with formatters and formatting lint rules and instead enforces the code is
```ts
() => {
  returnsVoid();
}
```

So in the interest of "satisfying a lint rule" people would need to add two (semantically) unnecessary lines that don't really add any clarity.

Though TBH I don't 100% agree with the option - but it was implemented [in the original TSLint rule](https://github.com/palantir/tslint/pull/2445) (2017) and thus [ported across when we rebuilt the rule](https://github.com/typescript-eslint/typescript-eslint/pull/2605) (2020). There's little actual discussion on why such an option was necessary though and TBH if we had our time again I'd probably push back on it to gather some concrete feedback and such before implementing it.

----

OTOH `foo = void voidFn()` is definitely **NOT** idiomatic JS. I spent [a while trying to find some cases in public github code](https://sourcegraph.com/search?q=context:global+/+%3D+void+%5B%5E0%7C%26%3E%3C%2Ce%28%5Cd%5D/+lang:TypeScript+-path:vendors%3F+-path:/%3F%28tests%3F%7Cout%7Cdist%7Cbuild%7Cdocs%29/&patternType=standard&sm=1&groupBy=repo) and I turned up with... not a whole lot.
The `void` operator is already rarely used in of itself.

I think for now this falls into the "too niche" bucket.
You're more than welcome to fork the rule locally into your project and implement it (or publish it as a separate plugin) - we encourage it!

If we find others asking for this functionality we can revisit this - but for now I just don't see the broad applicability

**Comment by Sxxov** [comment]
Thank you for the detailed explanation, further research, & simply everything you guys do for the TS ecosystem! Bit sad that this won't be fixed, but very happy with the strong sense of direction for the project.
